module/zfs/zvol.c

```diff
@@ -412,14 +412,11 @@ zvol_set_volsize(const char *name, uint64_t volsize)
 		os = zv->zv_objset;
 	}
 
-	doi = kmem_alloc(sizeof (dmu_object_info_t), KM_SLEEP);
-
-	if ((error = dmu_object_info(os, ZVOL_OBJ, doi)) ||
-	    (error = zvol_check_volsize(volsize, doi->doi_data_block_size)))
+	if ((error = dmu_object_info(os, ZVOL_OBJ, &doi)) ||
+	    (error = zvol_check_volsize(volsize, doi.doi_data_block_size)))
 		goto out;
 
 	error = zvol_update_volsize(volsize, os);
-	kmem_free(doi, sizeof (dmu_object_info_t));
```

Comment:
The concern here is to keep the stack size to a minimum.  Some older supported kernels still default to 8k stacks which we need to be mindful of.   We also need to be careful to keep individual stack frames small.  There are a few places in the code where these restrictions have forced us to move some larger structures off the stack.  I'm not sure what the worst case stack use is here, but I'd also suggest we take the conservative approach and move the `kmem_free()`.